snuba/datasets/entities/clickhouse_upgrade.py

```diff
@@ -94,3 +117,179 @@ def choose(self, referrer: str) -> Choice:
                 if not trust_secondary
                 else self.__default_primary,
             )
+
+
+Key = Tuple[Union[str, datetime, UUID], ...]
+RowValues = Tuple[Union[int, float, bool], ...]
+
+
+def comparison_callback(
+    _query: Query,
+    _settings: RequestSettings,
+    referrer: str,
+    primary_result: Optional[Result[QueryResult]],
+    results: List[Result[QueryResult]],
+) -> None:
+    """
+    Compares the primary and secondary results and logs metrics about the similarity.
+
+    The similarity is calculated fairly naively this way relying on the fact that
+    most queries have a number of string/datetime fields and a numebr of decimal
+    fields (int or float).
+
+    - We match the rows that are present in both results (matching the non decimal
+    fields)
+    - The similarity score is the average of the similarity of each number column.
+      Ex. row1 : (a, 10, 100, 1000) row2 (a, 9, 99, 990). The similarity between
+      these two is the avg([(smaller value / bigger value) for each column])
+    - Calculate the average of the similarity scores of all the matching rows.
+    - Rows that do not have a match have a similarity score = 0.0
+    """
+    if primary_result is None or len(results) == 0:
```

Comment:
It might be worth checking to make sure the secondary result also has 0 rows?

Reply:
`len(results) == 0` means there are no secondary results and we ran only one query.